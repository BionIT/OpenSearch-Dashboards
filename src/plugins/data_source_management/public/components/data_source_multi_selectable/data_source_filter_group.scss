.dataSourceFilterGroupItems {
  max-height: 400px;
  overflow: scroll;
  max-width: 300px;

  .dataSourceFilterGroupOuiFlexGroup {
    .dataSourceFilterGroupOuiFlexItem {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      display: inline-block;
    }
  }
}
