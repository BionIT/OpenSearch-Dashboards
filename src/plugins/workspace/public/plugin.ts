/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  createOsdUrlStateStorage,
  createStateContainer,
  syncState,
} from '../../opensearch_dashboards_utils/public';
import { CoreStart, Plugin } from '../../../core/public';

export class WorkspacePlugin implements Plugin<{}, {}, {}> {
  public async setup() {
    return {};
  }

  public start(core: CoreStart) {
    const initialState = {
      id: 'foo',
    };
    const globalQueryStateContainer = createStateContainer({});
    const urlStateStorageChange = createOsdUrlStateStorage({
      history: core.application.history,
      useHash: false,
    });
    const { start } = syncState({
      /**
       * Hijack urlStateStorage to call set with replace: true
       */
      stateStorage: {
        ...urlStateStorageChange,
        set(key, state) {
          return urlStateStorageChange.set(key, state, {
            replace: true,
          });
        },
      },
      stateContainer: {
        ...globalQueryStateContainer,
        set: (state) => {
          globalQueryStateContainer.set({
            ...initialState,
            ...state,
          });
        },
        get: () => {
          /**
           * Adding a timestamp field to make each response non-idempotent.
           * Or the syncState will compare each state, if no change it won't trigger stateContainer.set
           */
          return {
            ...initialState,
            timestamp: Date.now(),
          };
        },
      },
      storageKey: '_w',
    });
    start();
    return {};
  }

  public stop() {}
}
