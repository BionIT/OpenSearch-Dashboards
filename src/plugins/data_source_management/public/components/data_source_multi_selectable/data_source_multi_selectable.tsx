/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { FilterChecked } from '@elastic/eui';
import { SavedObjectsClientContract, ToastsStart } from 'opensearch-dashboards/public';
import { i18n } from '@osd/i18n';
import { DataSourceOption } from '../data_source_selector/data_source_selector';
import { FilterGroup } from './filter_group';
import { getDataSourcesWithFields } from '../utils';

export interface FilterItem {
  name: string | React.ReactNode;
  id: string;
  visible: boolean;
  childOptionIds?: Set<string>;
  checked?: FilterChecked;
}

export interface DataSourceMultiSeletableProps {
  savedObjectsClient: SavedObjectsClientContract;
  notifications: ToastsStart;
  onSelectedDataSources: (dataSources: DataSourceOption[] | FilterItem[]) => void;
  hideLocalCluster: boolean;
  fullWidth: boolean;
}

interface DataSourceMultiSeletableState {
  dataSourceOptions: DataSourceOption[] | FilterItem[];
  selectedOptions: DataSourceOption[] | FilterItem[];
}

export class DataSourceMultiSelectable extends React.Component<
  DataSourceMultiSeletableProps,
  DataSourceMultiSeletableState
> {
  private _isMounted: boolean = false;

  constructor(props: DataSourceMultiSeletableProps) {
    super(props);

    this.state = {
      dataSourceOptions: [],
      selectedOptions: [],
    };
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  async componentDidMount() {
    this._isMounted = true;
    getDataSourcesWithFields(this.props.savedObjectsClient, ['id', 'title', 'auth.type'])
      .then((fetchedDataSources) => {
        if (fetchedDataSources?.length) {
          const selectedOptions: FilterItem[] = fetchedDataSources.map((dataSource) => ({
            id: dataSource.id,
            name: dataSource.attributes?.title || '',
            checked: 'on',
            visible: true,
          }));

          if (!this.props.hideLocalCluster) {
            selectedOptions.unshift({
              id: '',
              name: 'Local cluster',
              checked: 'on',
              visible: true,
            });
          }

          if (!this._isMounted) return;
          this.setState({
            ...this.state,
            selectedOptions,
          });
        }
      })
      .catch(() => {
        this.props.notifications.addWarning(
          i18n.translate('dataSource.fetchDataSourceError', {
            defaultMessage: 'Unable to fetch existing data sources',
          })
        );
      });
  }

  onChange(items: FilterItem[]) {
    if (!this._isMounted) return;
    this.setState({
      selectedOptions: items,
    });
    this.props.onSelectedDataSources(items);
  }

  render() {
    return (
      <FilterGroup
        items={this.state.selectedOptions as FilterItem[]}
        hasGroupOptions={false}
        setItems={this.onChange.bind(this)}
      />
    );
  }
}
