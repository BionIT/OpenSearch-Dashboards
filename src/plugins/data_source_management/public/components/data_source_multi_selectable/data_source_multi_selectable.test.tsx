/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SavedObjectsClientContract } from 'opensearch-dashboards/public';
import { notificationServiceMock } from '../../../../../core/public/mocks';
import { getDataSourcesWithFieldsResponse, mockResponseForSavedObjectsCalls } from '../../mocks';
import { ShallowWrapper, shallow } from 'enzyme';
import { DataSourceMultiSelectable } from './data_source_multi_selectable';
import React from 'react';

describe('DataSourceMultiSelectable', () => {
  let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

  let client: SavedObjectsClientContract;
  const { toasts } = notificationServiceMock.createStartContract();

  beforeEach(() => {
    client = {
      find: jest.fn().mockResolvedValue([]),
    } as any;
    mockResponseForSavedObjectsCalls(client, 'find', getDataSourcesWithFieldsResponse);
  });

  it('should render normally with local cluster not hidden', () => {
    component = shallow(
      <DataSourceMultiSelectable
        savedObjectsClient={client}
        notifications={toasts}
        onSelectedDataSources={jest.fn()}
        hideLocalCluster={false}
        fullWidth={false}
      />
    );
    expect(component).toMatchSnapshot();
    expect(client.find).toBeCalledWith({
      fields: ['id', 'title', 'auth.type'],
      perPage: 10000,
      type: 'data-source',
    });
    expect(toasts.addWarning).toBeCalledTimes(0);
  });

  it('should render normally with local cluster hidden', () => {
    component = shallow(
      <DataSourceMultiSelectable
        savedObjectsClient={client}
        notifications={toasts}
        onSelectedDataSources={jest.fn()}
        hideLocalCluster={true}
        fullWidth={false}
      />
    );
    expect(component).toMatchSnapshot();
    expect(client.find).toBeCalledWith({
      fields: ['id', 'title', 'auth.type'],
      perPage: 10000,
      type: 'data-source',
    });
    expect(toasts.addWarning).toBeCalledTimes(0);
  });
});
