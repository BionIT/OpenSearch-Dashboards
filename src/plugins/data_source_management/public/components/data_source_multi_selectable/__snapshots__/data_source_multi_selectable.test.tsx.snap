// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceMultiSelectable should render normally with local cluster hidden 1`] = `
<FilterGroup
  hasGroupOptions={false}
  items={Array []}
  setItems={[Function]}
/>
`;

exports[`DataSourceMultiSelectable should render normally with local cluster not hidden 1`] = `
<FilterGroup
  hasGroupOptions={false}
  items={Array []}
  setItems={[Function]}
/>
`;
